README.md
# openhab-cookbook

This installs [openhab](http://www.openhab.org/) and selected plugins.  You can either install via the debian packages, or from source.

NOTE: This is only for versions 1.x of openhab.  When I move to OpenHAB 2, I'll start a `openhab2` cookbook.

### Dependencies

- Cookbooks:
  - apt
  - java
- On the system:
 - apt updated (including `apt` in the run-list will do this)
 - java 8 installed
   - this cookbook will install java for you by default - make sure to set the attributes for oracle v8

## Supported Platforms

Tested succesful on:

- Ubuntu 12.04
- Ubuntu 14.04
- Raspbian (source-only)

## Recipies

- `openhab` - Installs openhab
- `openhab::init_script` - Installs openhab init.d script
- `openhab::my_openhab`  - Installs my.openhab addon .jar
- `openhab::habmin` - Installs [HABmin console](https://github.com/cdjackson/HABmin)

## Attributes

### Global

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['openhab']['install_java']</tt></td>
    <td>Boolean</td>
    <td>Whether to install java via the java cookbook</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['install_method']</tt></td>
    <td>String</td>
    <td>Whether to install openhab via the deb packages (default) or via 'source'</td>
    <td><tt>'package'</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['version']</tt></td>
    <td>String</td>
    <td>What version of openhab to install</td>
    <td><tt>1.7.1</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['plugins']</tt></td>
    <td>Array</td>
    <td>List of plugins to install</td>
    <td><tt>[]</tt></td>
  </tr>
</table>

### Init script install only

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['openhab']['http_port']</tt></td>
    <td>String</td>
    <td>HTTP Port for the init script</td>
    <td><tt>8080</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['https_port']</tt></td>
    <td>String</td>
    <td>HTTPS Port for the init script</td>
    <td><tt>8443</tt></td>
  </tr>
</table>

### Source install only

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['openhab']['install_dir']</tt></td>
    <td>String</td>
    <td>Which directoy to install 'openhab' into</td>
    <td><tt>'/opt'</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['user']</tt></td>
    <td>String</td>
    <td>What user to run install as</td>
    <td><tt>root</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['group']</tt></td>
    <td>String</td>
    <td>What user to run install as</td>
    <td><tt>root</tt></td>
  </tr>
</table>

### my.openHAB plugin

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['openhab']['my']['version']</tt></td>
    <td>String</td>
    <td>Which version of the my.openhab plugin to install</td>
    <td><tt>1.7.0</tt></td>
  </tr>
</table>

### habmin console

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['openhab']['habmin']['install_dir']</tt></td>
    <td>String</td>
    <td>Where to install habmin</td>
    <td><tt>/opt/openhab/webapps/habmin</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['habmin']['git_repository']</tt></td>
    <td>String</td>
    <td>Which repo to download habmin from</td>
    <td><tt>https://github.com/cdjackson/HABmin</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['habmin']['git_revision']</tt></td>
    <td>String</td>
    <td>Which ref of habmin to use</td>
    <td><tt>e523d77c7a3d73d75ce050a582a3bfe8bc7a1324</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['habmin']['user']</tt></td>
    <td>String</td>
    <td>What user to run install as</td>
    <td><tt>root</tt></td>
  </tr>
  <tr>
    <td><tt>['openhab']['habmin']['group']</tt></td>
    <td>String</td>
    <td>What group to run install as</td>
    <td><tt>root</tt></td>
  </tr>
</table>

## Usage

- 1. Make sure you have Java setup:

```ruby
{
  attributes: {
    java: {
      install_flavor: 'oracle',
      jdk_version: '8',
      oracle: {accept_oracle_download_terms: true}
    }
  }
}
```

- 2. Pick some plugins:

```ruby
{
  attributes: {
    openhab: {
      plugins: %w{binding-xbmc action-xbmc binding-twitter persistance-mysql}
    }
  }
}
```

- 3. Include `openhab` in your node's run_list:

```ruby
{
  "run_list": [
    "recipe[apt]",
    "recipe[openhab]"
  ]
}
```

- 4. ???
- 5. Profit!

You now have OpenHAB installed.  If you did it from the packages, you have the init.d script set up and openhab running already.

### Init script

If you want to install an init.d script for booting, use the `openhab::init_script` recipie.  Currently only Raspbian is supported.

### my.openhab plugin

If you want to install the myOpenHABplugin, run the `openhab::my_openhab` recipie - it will place the requested .jar in the addons dir.

You'll still need to set/get the uuid/secret registration yourself.

## License and Authors

Author:: [Justin Aiken](https://github.com/JustinAiken)


CHANGELOG.md
# 0.0.4
- Add recipe to install [HABmin](https://github.com/cdjackson/HABmin)

# 0.0.3
- Support 1.7.x releases, now hosted on bintray
- Support 1.7.x myopenhab releases, now hosted on bintray
- [BUGFIX] Fix some directories being changed w/ ark by setting `strip_components` to `0`
- [BUGFIX] Fix a bad plugin regex

# 0.0.2
- Add raspbian init.d script
- Add myopenhab plugin

# 0.0.1

Initial release of openhab


test/integration/from_source/serverspec/openhab_spec.rb
require 'serverspec'

# Required by serverspec
set :backend, :exec

describe 'openhab' do
  it "creates the directory" do
    expect(file '/opt/openhab').to be_directory
  end

  it "has some enabled plugins" do
    expect(file '/opt/openhab/addons/org.openhab.action.xbmc-1.6.2.jar').to be_file
    expect(file '/opt/openhab/addons/org.openhab.binding.xbmc-1.6.2.jar').to be_file
  end

  it "has some disabled plugins" do
    expect(file '/opt/openhab/disabled-addons/org.openhab.binding.mqtt-1.6.2.jar').to be_file
  end
end


test/integration/from_package/serverspec/openhab_spec.rb
require 'serverspec'

# Required by serverspec
set :backend, :exec

describe 'openhab' do
  it "installs the openhab package" do
    expect(package 'openhab-runtime').to be_installed
  end

  it "installs the requested plugins" do
    expect(package 'openhab-addon-action-xbmc').to  be_installed
    expect(package 'openhab-addon-binding-xbmc').to be_installed
  end

  it "serves" do
    expect(service :openhab).to be_running
    expect(port 8080).to be_listening
  end
end


recipes/my_openhab.rb
#
# Cookbook Name:: openhab
# Recipe:: my_openhab
#
# Copyright (C) 2015 Justin Aiken
#
# MIT
#

# %w{webapps webapps/static}.each do |dir|
#   directory "#{node[:openhab][:install_dir]}/openhab/#{dir}" do
#     owner node[:openhab][:user]
#     group node[:openhab][:group]
#     action :create
#   end
# end

if node[:openhab][:my][:version] < "1.7.0"
  JAR_NAME   = "org.openhab.io.myopenhab-#{node[:openhab][:my][:version]}-SNAPSHOT.jar"
  SOURCE_URL = "https://my.openhab.org/downloads/#{JAR_NAME}"
else
  JAR_NAME   = "org.openhab.io.myopenhab-#{node[:openhab][:my][:version]}.jar"
  SOURCE_URL = "https://bintray.com/artifact/download/openhab/mvn/org/openhab/io/org.openhab.io.myopenhab/#{node[:openhab][:my][:version]}/#{JAR_NAME}"
end

remote_file "#{node[:openhab][:install_dir]}/openhab/addons/#{JAR_NAME}" do
  source SOURCE_URL
  owner node[:openhab][:user]
  group node[:openhab][:group]
  action :create_if_missing
end


recipes/install_from_source.rb
#
# Cookbook Name:: openhab
# Recipe:: install_from_source
#
# Copyright (C) 2015 Justin Aiken
#
# MIT
#

if node[:openhab][:version] < "1.6.2"
  RELEASE_URL = "https://github.com/openhab/openhab/releases/download/"
  VERSION_URL = "#{RELEASE_URL}v#{node[:openhab][:version]}/distribution-#{node[:openhab][:version]}-"
else
  RELEASE_URL = "https://bintray.com/artifact/download/openhab/bin/"
  VERSION_URL = "#{RELEASE_URL}distribution-#{node[:openhab][:version]}-"
end

package 'unzip'

ark "openhab-#{node[:openhab][:version]}" do
  name 'openhab'
  url "#{VERSION_URL}runtime.zip"
  version node[:openhab][:version]
  path node[:openhab][:install_dir]
  owner node[:openhab][:user]
  group node[:openhab][:group]
  strip_components 0
  action :put
end

PLUGIN_REGEX = /\A(action|binding|persistance|io)\.(.*)\z/

if node['openhab']['plugins'].count > 0

  ark 'openhab-addons' do
    url "#{VERSION_URL}addons.zip"
    version node[:openhab][:version]
    path "#{node[:openhab][:install_dir]}/openhab/disabled-addons"
    owner node[:openhab][:user]
    group node[:openhab][:group]

    action :dump
  end

  directory "#{node[:openhab][:install_dir]}/openhab/addons" do
    owner node[:openhab][:user]
    group node[:openhab][:group]
    action :create
  end

  node['openhab']['plugins'].each do |plugin|
    jar_name = plugin.gsub(PLUGIN_REGEX, 'org.openhab.\1.\2') + "-#{node[:openhab][:version]}.jar"

    bash "install #{jar_name}" do
      cwd  "#{node[:openhab][:install_dir]}/openhab/addons"
      code "mv ../disabled-addons/#{jar_name} ."
      only_if { ::File.exists? "#{node[:openhab][:install_dir]}/openhab/disabled-addons/#{jar_name}" }
      not_if  { ::File.exists? "#{node[:openhab][:install_dir]}/openhab/addons/#{jar_name}"          }
    end
  end
end


attributes/default.rb
#
# Cookbook Name:: openhab
# Attributes:: default
#
# Copyright 2015, Justin Aiken
#

default['openhab']['install_java']   = true
default['openhab']['install_method'] = 'package'
default['openhab']['version']        = '1.7.1'
default['openhab']['plugins']        = []

# Source only options:
default['openhab']['install_dir'] = '/opt'
default['openhab']['user']        = 'root'
default['openhab']['group']       = 'root'

# Init script options:
default['openhab']['http_port']  = '8080'
default['openhab']['https_port'] = '8443'

# my.openHAB options:
default['openhab']['my']['version'] = '1.7.0'

# habmin options:
default['openhab']['habmin']['install_dir']    = '/opt/openhab/webapps/habmin'
default['openhab']['habmin']['git_repository'] = 'https://github.com/cdjackson/HABmin'
default['openhab']['habmin']['git_revision']   = 'e523d77c7a3d73d75ce050a582a3bfe8bc7a1324' # v0.1.3-snapshot
default['openhab']['habmin']['user']           = 'root'
default['openhab']['habmin']['group']          = 'root'
